### ISSUE TITLE
RatbagdButton.Type always returns unknown on at least the G403

### ISSUE THREAD

**Issue by Hjdskes** [libratbag/libratbag#233]
Noticed while implementing button mappings in Piper: the `Type` property on a `RatbagdButton` DBus object always returns `unknown`. [According to](https://github.com/libratbag/piper/pull/47#issuecomment-315944951) @whot this is a bug, hence this issue.

**Comment by bentiss** [comment]
Note that `type` refers to the button physical type (left, right, thumb, etc...) not the associated binding. It's read only (given that it's hardware dependent) and is a per-driver bug, not a general one.

**Comment by Hjdskes** [comment]
Yes, I am aware of that. In Piper it's needed to display the function of a button mapping, i.e. if I assign logical button 1 to physical button 0, in Piper it should say "Right mouse button click".

**Comment by bentiss** [comment]
> if I assign logical button 1 to physical button 0, in Piper it should say "Right mouse button click".

I still think you are confused here. The `type` is the position of the button on/under the mouse. It has nothing to do with bindings. For instance on the G700, I have 4 thumbs buttons and they are typed as `thumb`, `thumb2`, `thumb3` and `thumb4`. On piper, there is absolutely no need to retrieve this property given you have the SVG to rely on which is far more explicit than any type we could think of.

**Comment by Hjdskes** [comment]
I don't think I am confused. If I map to my physical button 0 (i.e. the left mouse button) the logical button 1 (i.e. the right mouse button), then in Piper it should display "Right mouse button click". In order to do so, I need to know the type of the logical button that my physical button is mapped to, so that I can grab the correct string in the type-to-description table [here](https://github.com/libratbag/piper/pull/47/files#diff-0cbd45bbfc400e5b65a19e03dd9dcca4R450).

How would I do this without needing the logical button's type?

**Comment by bentiss** [comment]
The fact that you name `button 1` to `Right mouse button click` has nothing to do with ratbagd and further more the `type` property of a button. It's a convention in X that button 1 is the right click, but you should not use the physical information from `type` to get to the description of what a button does.

Also the fact that logical button 1 manages to map to Right click (which should not be working given that right click is at the third position) would be pure coincidence.

Again, `type` is the physical description of a physical button and it is not supposed to change during the lifetime of the device. If you want to have fancy naming (button 1 to right click, button 2 to middle, 14 to forward IIRC, etc...), then this is something piper specific and it should not end up in ratbagd.py

**Comment by Hjdskes** [comment]
I am indeed confusing some things. It was @whot's [comment](https://github.com/libratbag/piper/pull/47#issuecomment-315944951):

> I think this is something where the GUI has to make the decisions and limit what can be done compared to the raw interface on the commandline. It's acceptable to have a list of "left click", "right click", ... etc and then map those to the respective button numbers without having to expose the button number option to the user. The wacom configuration is a guideline here, see https://wiki.gnome.org/Design/SystemSettings/Tablet though that one is not up to date. Look at the stylus config, it says "Left button click". I think we should do the same here.

that made me think that this was the right approach, i.e. "map those to the respective button numbers". @cvuchener's [comment](https://github.com/libratbag/piper/pull/47#issuecomment-316040785) now also makes a lot more sense. 

Would an alternative, correct approach be to have a list within Piper of the following:

button 0 → left click
button 1 → right click
button 2 → middle click
button 3 → back
button 4 → forward

as suggested by @cvuchener in the linked comment, and leave the rest as-is? (e.g. "button 5", "button 6", et cetera)

**Comment by bentiss** [comment]
Yes, this would be much better.

**Comment by bentiss** [comment]
Couple of things:

- keeping this one still opened as there are some work to be done in the drivers to have proper `type` support

- I think we introduced the `type` of a button to allow the UI to restrict the list of choices when it sees a button of a certain type. It _could_ be interesting to implement in piper "physical Left and Right buttons can only be remapped to button 0 and 1". This adds restrictions on the usage, but we could also make some sanity check on the mappings when there is something obviously wrong ("you have 2 profiles enabled and can't change them on the mouse", "you have no primary or secondary click, are you using it as a mouse?", etc...)

**Comment by whot** [comment]
@Hjdskes merged the patch for the left/right button swap in piper, see above.

What are we going to do with this one now? As a temporary fix we could just add this to the `.device` files until we know how to coerce the driver into giving us this info
